compiler/rustc_const_eval/src/util/might_permit_raw_init.rs

```diff
@@ -108,7 +108,11 @@ fn might_permit_raw_init_lax<'tcx>(
 
     // Special magic check for references and boxes (i.e., special pointer types).
     if let Some(pointee) = this.ty.builtin_deref(false) {
-        let pointee = cx.layout_of(pointee.ty).expect("need to be able to compute layouts");
+        let Ok(pointee) = cx.layout_of(pointee.ty) else {
+            // Reference is too polymorphic, it has a layout but the pointee does not.
+            assert!(pointee.ty.needs_subst());
+            return false;
```

Comment:
oh, lol, this is totally wrong. this is *`might_permit_raw_init_lax`*, not `does_permit_raw_init_lax` :/ this probably should return `true`...

@saethlin, are you sure that it's valid to instcombine away an uninit validity assertion in this case? really we should be returning something ternary -- yes, no, or "too generic" (or at least, "unknown layout"), and in the case of the last option, we don't do anything in the optimization.

Reply:
Not sure what you're asking, but I'll guess.

If you are asking me what the correct thing to do instead of ICEing here is, no idea. The upstream effect here should be not applying the optimization- we don't know if `T` has alignment > 1, in which case the 0x1-filling is invalid.

If you're asking about the code I added to InstCombine, I intended to nab that directly from what codegen does when it decides to emit a panic or not for that intrinsic. Now it's a fair question as to whether what codegen does just behaves nonsensically when given a generic, if code is supposed to be monomorphized at that stage.

Reply:
What I'm saying is that it's probably not correct to return either `false` (like I did) or `true` if we can't get the layout of the pointee in this function. 

[CI failure](https://github.com/rust-lang/rust/pull/108012#issuecomment-1428863945) shows that `false` is bad because we're optimizing away perfectly fine polymorphic functions, and `true` is also probably not great, since we're dropping this assertion which may be false later once things are monomorphized. Added a mir-opt test to show that behavior in the first commit in this PR.

Second PR in this stack makes the queries you added fallible, so we can handle the `LayoutError`s gracefully and not perform the optimization in that case. It duplicates a bit of work, since we may have to call the layout query twice on some code paths, but oh well.

Reply:
> and `true` is also probably not great, since we're dropping this assertion

If this function is not allowed to have false positives maybe it shouldn't be called `might_`

Reply:
The function is intended to definitely answer whether we will emit a panic or not. Optimizations shouldn't change that behavior.

We don't always emit panics fully aggressively (that depends on compiler flags), so in a sense there are false negatives allowed.